pkg/agent/openflow/pipeline.go

```diff
@@ -1890,6 +2085,114 @@ func (c *client) snatRuleFlow(ofPort uint32, snatIP net.IP, snatMark uint32, loc
 	}
 }
 
+func (c *client) initServiceClassifierFlows(nodePortIPMap map[int][]net.IP, isIPv6 bool) []binding.Flow {
+	clusterConjID := clusterConjIDIPv4
+	localConjID := localConjIDIPv4
+	ipProtocol := binding.ProtocolIP
+	serviceGWHairpinIP := config.ServiceGWHairpinIPv4
+	if isIPv6 {
+		clusterConjID = clusterConjDIPv6
+		localConjID = localConjIDIPv6
+		ipProtocol = binding.ProtocolIPv6
+		serviceGWHairpinIP = config.ServiceGWHairpinIPv6
+	}
+
+	var flows []binding.Flow
+	for _, ips := range nodePortIPMap {
+		for _, ip := range ips {
+			flows = append(flows,
+				// This flow is used to match the first packet's destination IP address:
+				// 1. NodePort Service whose externalTrafficPolicy is Cluster, and client is from remote/localhost.
+				// 2. NodePort Service whose externalTrafficPolicy is Local, and client is from remote.
+				c.pipeline[serviceClassifierTable].BuildFlow(priorityNormal).
+					Cookie(c.cookieAllocator.Request(cookie.Service).Raw()).
+					MatchProtocol(ipProtocol).
+					MatchDstIP(ip).
+					Action().Conjunction(clusterConjID, 1, 2).
```

Comment:
I'm thinking if it really needs an extra serviceClassifierTable to handle NodePort/LoadBalancer traffic.
What the table does is to mark whether this is a NodePort traffic and whether it needs SNAT:
For whether this is a NodePort traffic, what if we just match the service virtual IP? The NodePort traffic from local Pods will be output to antrea-gw0 and get DNATed there and sent back to OVS, everything should just work. There should be no problem even the endpoint is a local Pod because `gatewayCTMark` related flow guarantees the reply packets will be sent back to antrea-gw0.
Given that accessing NodePort from Pod directly is not a typical scenario and the extra flows only handle local NodeIP but not all Node IPs, I think we can just leave the traffic to host netns.

For whether it needs SNAT, I think it can be done in `serviceLBTable`, we can just pass `nodeLocalExternal` to `InstallServiceFlows` and load the mark based on it before the `Group` action.

Then `serviceClassifierTable` is not needed at all and all group selection logic is done in `serviceLBTable`.

Reply:
You are right. It can work as you say. The `serviceClassifierTable` is not needed at all. `serviceLBTable` can take over the work of writing SNAT register.